xarray/core/dataarray.py

```diff
@@ -130,18 +130,30 @@
     T_XarrayOther = TypeVar("T_XarrayOther", bound="DataArray" | Dataset)
 
 
-def _check_coords_dims(shape, coords, dim):
+def _check_coords_dims(
+    shape: tuple[int, ...], coords: Coordinates, dim: tuple[Hashable, ...]
+):
     sizes = dict(zip(dim, shape, strict=True))
+    extra_index_dims: set[str] = set()
+
     for k, v in coords.items():
         if any(d not in dim for d in v.dims):
+            # allow any coordinate associated with an index that shares at least
+            # one of dataarray's dimensions
+            indexes = coords.xindexes
+            if k in indexes:
+                index_dims = indexes.get_all_dims(k)
+                if any(d in dim for d in index_dims):
+                    extra_index_dims.update(d for d in v.dims if d not in dim)
+                    continue
             raise ValueError(
                 f"coordinate {k} has dimensions {v.dims}, but these "
                 "are not a subset of the DataArray "
                 f"dimensions {dim}"
             )
 
         for d, s in v.sizes.items():
-            if s != sizes[d]:
+            if d not in extra_index_dims and s != sizes[d]:
```

Comment:
```suggestion
            if d not in extra_index_dims or s != sizes[d]:
```
?

Reply:
This should be `and`.

`extra_index_dims` corresponds to all non-array dimensions of index coordinates to include in the dataarray (size[d] would return a KeyError).

I renamed it and added comment in [695fb86](https://github.com/pydata/xarray/pull/10116/commits/695fb8611c16d64a4f1b199e1324ecfc53592c5e).